File: sdk/python/src/screenmcp/selector.py
"""Selector engine for finding UI elements by query."""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any


@dataclass
class FoundElement:
    """A UI element found by a selector query."""

    x: int
    y: int
    bounds: dict
    text: str | None = None
    class_name: str | None = None
    resource_id: str | None = None
    content_description: str | None = None
    node: dict = field(default_factory=dict)


# Maps selector prefix to (node field, operation)
_FIELD_MAP = {
    "text": "text",
    "role": "className",
    "desc": "contentDescription",
    "id": "resourceId",
}

# Regex for a single selector atom: optional !, prefix:value or prefix=value,
# optional property filters [prop], optional index [n]
_ATOM_RE = re.compile(
    r"(?P<neg>!)?"
    r"(?P<prefix>text|role|desc|id)"
    r"(?P<op>[:=])"
    r"(?P<value>[^\[]+?)"
    r"(?P<filters>(?:\[[a-zA-Z_]+\])*)"
    r"(?:\[(?P<index>-?\d+)\])?"
    r"$"
)


def parse_selector(selector: str) -> list[dict]:
    """Parse selector string into query structure (list of OR groups).

    Each OR group is a list of condition dicts (AND-ed together).
    Returns a list of OR groups.
    """
    or_groups = [g.strip() for g in selector.split("||")]
    parsed: list[dict] = []

    for group in or_groups:
        and_parts = [p.strip() for p in group.split("&&")]
        conditions: list[dict[str, Any]] = []
        group_index: int | None = None

        for part in and_parts:
            m = _ATOM_RE.match(part)
            if not m:
                raise ValueError(f"Invalid selector: {part!r}")

            prefix = m.group("prefix")
            op = "equals" if m.group("op") == "=" else "contains"
            value = m.group("value")
            negated = m.group("neg") == "!"

            node_field = _FIELD_MAP[prefix]

            # Parse property filters like [focused][clickable]
            prop_filters: list[str] = []
            if m.group("filters"):
                prop_filters = re.findall(r"\[([a-zA-Z_]+)\]", m.group("filters"))

            # Parse index
            if m.group("index") is not None:
                group_index = int(m.group("index"))

            conditions.append({
                "field": node_field,
                "value": value,
                "op": op,
                "negated": negated,
                "prop_filters": prop_filters,
                "is_role": prefix == "role",
            })

        parsed.append({
            "conditions": conditions,
            "index": group_index,
        })

    return parsed


def match_node(node: dict, query: dict) -> bool:
    """Check if a single node matches a parsed query group.

    ``query`` is one element from the list returned by ``parse_selector``.
    Only the conditions are checked here; index filtering is done externally.
    """
    for cond in query["conditions"]:
        node_field: str = cond["field"]
        target: str = cond["value"]
        op: str = cond["op"]
        negated: bool = cond["negated"]
        prop_filters: list[str] = cond["prop_filters"]
        is_role: bool = cond["is_role"]

        # Get the field value from the node.
        # For role selectors, also check 'title' (desktop nodes).
        node_value = node.get(node_field)
        if is_role and node_value is None:
            node_value = node.get("title")

        # Perform the match
        if op == "equals":
            matched = node_value is not None and str(node_value) == target
        else:  # contains, case-insensitive
            matched = (
                node_value is not None
                and target.lower() in str(node_value).lower()
            )

        if negated:
            matched = not matched

        if not matched:
            return False

        # Check property filters
        for prop in prop_filters:
            if not node.get(prop):
                return False

    return True


def _compute_bounds(node: dict) -> tuple[int, int, dict] | None:
    """Compute center coordinates and bounds dict from a node.

    Returns (center_x, center_y, bounds_dict) or None if no geometry.
    """
    bounds = node.get("bounds")
    if bounds and "left" in bounds:
        # Android-style bounds
        left = bounds["left"]
        top = bounds["top"]
        right = bounds["right"]
        bottom = bounds["bottom"]
        cx = (left + right) // 2
        cy = (top + bottom) // 2
        return cx, cy, {"left": left, "top": top, "right": right, "bottom": bottom}

    # Desktop-style bounds
    if "x" in node and "y" in node and "width" in node and "height" in node:
        x = node["x"]
        y = node["y"]
        w = node["width"]
        h = node["height"]
        cx = x + w // 2
        cy = y + h // 2
        return cx, cy, {"left": x, "top": y, "right": x + w, "bottom": y + h}

    return None


def _walk_tree(tree: list | dict, query: dict) -> list[FoundElement]:
    """Recursively walk the tree and collect matching elements."""
    results: list[FoundElement] = []

    if isinstance(tree, dict):
        nodes = [tree]
    elif isinstance(tree, list):
        nodes = tree
    else:
        return results

    for node in nodes:
        if match_node(node, query):
            geo = _compute_bounds(node)
            if geo is not None:
                cx, cy, bounds = geo
                results.append(FoundElement(
                    x=cx,
                    y=cy,
                    bounds=bounds,
                    text=node.get("text"),
                    class_name=node.get("className"),
                    resource_id=node.get("resourceId"),
                    content_description=node.get("contentDescription"),
                    node=node,
                ))

        # Recurse into children
        children = node.get("children")
        if children:
            results.extend(_walk_tree(children, query))

    return results


def find_elements(tree: list, selector: str) -> list[FoundElement]:
    """Walk tree recursively, return all matching nodes.

    Supports OR groups (``||``), AND conditions (``&&``),
    negation (``!``), property filters (``[focused]``),
    and index selection (``[0]``, ``[-1]``).
    """
    parsed = parse_selector(selector)
    all_results: list[FoundElement] = []

    for group in parsed:
        matches = _walk_tree(tree, group)
        idx = group.get("index")
        if idx is not None and matches:
            try:
                matches = [matches[idx]]
            except IndexError:
                matches = []
        all_results.extend(matches)

    return all_results


class ElementHandle:
    """Fluent handle for interacting with a found element."""

    def __init__(
        self, client: Any, selector: str, timeout: float = 3.0
    ) -> None:
        self._client = client
        self._selector = selector
        self._timeout = timeout

    async def _resolve(self) -> FoundElement:
        """Poll ui_tree until element is found or timeout."""
        import asyncio
        import time

        deadline = time.monotonic() + self._timeout
        while True:
            result = await self._client.ui_tree()
            tree = result.get("tree", [])
            found = find_elements(tree, self._selector)
            if found:
                return found[0]
            if time.monotonic() >= deadline:
                from .client import ScreenMCPError

                raise ScreenMCPError(f"Element not found: {self._selector}")
            await asyncio.sleep(0.5)

    async def click(self) -> dict[str, Any]:
        """Click the found element."""
        el = await self._resolve()
        return await self._client.click(el.x, el.y)

    async def long_click(self) -> dict[str, Any]:
        """Long-click the found element."""
        el = await self._resolve()
        return await self._client.long_click(el.x, el.y)

    async def type(self, text: str) -> dict[str, Any]:
        """Click the element to focus it, then type text."""
        import asyncio

        el = await self._resolve()
        await self._client.click(el.x, el.y)
        await asyncio.sleep(0.3)
        return await self._client.type_text(text)

    async def element(self) -> FoundElement:
        """Resolve and return the FoundElement."""
        return await self._resolve()

File: sdk/typescript/src/selector.ts
// ---------------------------------------------------------------------------
// Selector engine: parse selector strings and match against UI tree nodes
// ---------------------------------------------------------------------------

export type SelectorOp = "contains" | "equals";

export interface SelectorAtom {
  field: string; // 'text', 'className', 'contentDescription', 'resourceId'
  value: string;
  op: SelectorOp;
  negated: boolean;
}

export interface PropertyFilter {
  prop: string; // 'focused', 'clickable', 'checked', etc.
}

export interface ParsedSelector {
  atoms: SelectorAtom[];
  properties: PropertyFilter[];
  index?: number;
}

/** Top-level: OR groups */
export type SelectorQuery = ParsedSelector[];

export interface FoundElement {
  x: number;
  y: number;
  bounds: { left: number; top: number; right: number; bottom: number };
  text?: string;
  className?: string;
  resourceId?: string;
  contentDescription?: string;
  node: Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// Parser
// ---------------------------------------------------------------------------

const FIELD_MAP: Record<string, string> = {
  text: "text",
  role: "className",
  desc: "contentDescription",
  id: "resourceId",
};

function parseAtom(raw: string): {
  atom: SelectorAtom;
  rest: string;
} {
  const negated = raw.startsWith("!");
  const s = negated ? raw.slice(1) : raw;

  // Try field=value (equals)
  const eqIdx = s.indexOf("=");
  const colonIdx = s.indexOf(":");

  let field: string;
  let value: string;
  let op: SelectorOp;
  let consumed: number;

  if (eqIdx !== -1 && (colonIdx === -1 || eqIdx < colonIdx)) {
    const key = s.slice(0, eqIdx);
    field = FIELD_MAP[key] ?? key;
    op = "equals";
    // Value goes until && or || or [ or end
    const rest = s.slice(eqIdx + 1);
    const endIdx = findValueEnd(rest);
    value = rest.slice(0, endIdx);
    consumed = (negated ? 1 : 0) + eqIdx + 1 + endIdx;
  } else if (colonIdx !== -1) {
    const key = s.slice(0, colonIdx);
    field = FIELD_MAP[key] ?? key;
    op = "contains";
    const rest = s.slice(colonIdx + 1);
    const endIdx = findValueEnd(rest);
    value = rest.slice(0, endIdx);
    consumed = (negated ? 1 : 0) + colonIdx + 1 + endIdx;
  } else {
    throw new Error(`Invalid selector atom: ${raw}`);
  }

  return {
    atom: { field, value, op, negated },
    rest: raw.slice(consumed),
  };
}

/** Find where the value ends: at &&, ||, [, or end of string */
function findValueEnd(s: string): number {
  for (let i = 0; i < s.length; i++) {
    if (s[i] === "[") return i;
    if (s[i] === "&" && s[i + 1] === "&") return i;
    if (s[i] === "|" && s[i + 1] === "|") return i;
  }
  return s.length;
}

function parseBrackets(s: string): {
  properties: PropertyFilter[];
  index?: number;
  rest: string;
} {
  const properties: PropertyFilter[] = [];
  let index: number | undefined;
  let rest = s;

  while (rest.startsWith("[")) {
    const close = rest.indexOf("]");
    if (close === -1) throw new Error(`Unclosed bracket in selector: ${s}`);
    const inner = rest.slice(1, close).trim();
    rest = rest.slice(close + 1);

    // Check if it's a numeric index
    const num = Number(inner);
    if (!isNaN(num) && inner !== "") {
      index = num;
    } else {
      properties.push({ prop: inner });
    }
  }

  return { properties, index, rest };
}

function parseGroup(s: string): ParsedSelector {
  const atoms: SelectorAtom[] = [];
  let remaining = s.trim();

  while (remaining.length > 0) {
    // Skip whitespace around &&
    remaining = remaining.replace(/^\s*&&\s*/, "");
    if (remaining.length === 0) break;

    // If starts with [, parse brackets
    if (remaining.startsWith("[")) {
      const { properties, index, rest } = parseBrackets(remaining);
      const result: ParsedSelector = { atoms, properties, index };
      // There might be more after brackets
      if (rest.trim().length > 0) {
        const continued = parseGroup(rest.trim());
        result.atoms.push(...continued.atoms);
        result.properties.push(...continued.properties);
        if (continued.index !== undefined) result.index = continued.index;
      }
      return result;
    }

    const { atom, rest } = parseAtom(remaining);
    atoms.push(atom);
    remaining = rest;

    // Parse any brackets attached to this atom
    if (remaining.startsWith("[")) {
      const { properties, index, rest: afterBrackets } =
        parseBrackets(remaining);
      remaining = afterBrackets;
      // If there's more with &&, continue
      if (remaining.startsWith("&&")) {
        remaining = remaining.slice(2).trim();
        const continued = parseGroup(remaining);
        return {
          atoms: [...atoms, ...continued.atoms],
          properties: [...properties, ...continued.properties],
          index: index ?? continued.index,
        };
      }
      return { atoms, properties, index };
    }
  }

  return { atoms, properties: [] };
}

/** Parse a selector string into a query (array of OR groups). */
export function parseSelector(selector: string): SelectorQuery {
  const groups = selector.split("||").map((s) => s.trim());
  return groups.map(parseGroup);
}

// ---------------------------------------------------------------------------
// Matcher
// ---------------------------------------------------------------------------

function getNodeField(
  node: Record<string, unknown>,
  field: string,
): string | undefined {
  // Direct field lookup
  if (field in node && node[field] != null) {
    return String(node[field]);
  }
  // For 'className', also try 'title' (desktop nodes)
  if (field === "className" && "title" in node && node.title != null) {
    return String(node.title);
  }
  return undefined;
}

/** Test whether a single node matches a parsed selector group. */
export function matchNode(
  node: Record<string, unknown>,
  query: ParsedSelector,
): boolean {
  // Check all atoms (ANDed)
  for (const atom of query.atoms) {
    const fieldVal = getNodeField(node, atom.field);
    let matches: boolean;

    if (fieldVal === undefined) {
      matches = false;
    } else if (atom.op === "equals") {
      matches = fieldVal === atom.value;
    } else {
      // contains, case-insensitive
      matches = fieldVal.toLowerCase().includes(atom.value.toLowerCase());
    }

    if (atom.negated) matches = !matches;
    if (!matches) return false;
  }

  // Check property filters (ANDed)
  for (const pf of query.properties) {
    if (!node[pf.prop]) return false;
  }

  return true;
}

function extractBounds(node: Record<string, unknown>): FoundElement | null {
  // Android format: bounds object
  const bounds = node.bounds as
    | { left: number; top: number; right: number; bottom: number }
    | undefined;
  if (bounds && typeof bounds.left === "number") {
    return {
      x: Math.round((bounds.left + bounds.right) / 2),
      y: Math.round((bounds.top + bounds.bottom) / 2),
      bounds,
      text: node.text as string | undefined,
      className: node.className as string | undefined,
      resourceId: node.resourceId as string | undefined,
      contentDescription: node.contentDescription as string | undefined,
      node,
    };
  }

  // Desktop format: x, y, width, height
  if (
    typeof node.x === "number" &&
    typeof node.y === "number" &&
    typeof node.width === "number" &&
    typeof node.height === "number"
  ) {
    const left = node.x as number;
    const top = node.y as number;
    const width = node.width as number;
    const height = node.height as number;
    return {
      x: Math.round(left + width / 2),
      y: Math.round(top + height / 2),
      bounds: { left, top, right: left + width, bottom: top + height },
      text: node.text as string | undefined,
      className: (node.className ?? node.title) as string | undefined,
      resourceId: node.resourceId as string | undefined,
      contentDescription: node.contentDescription as string | undefined,
      node,
    };
  }

  return null;
}

function collectMatches(
  nodes: unknown[],
  query: ParsedSelector,
  results: FoundElement[],
): void {
  for (const raw of nodes) {
    if (typeof raw !== "object" || raw === null) continue;
    const node = raw as Record<string, unknown>;

    if (matchNode(node, query)) {
      const el = extractBounds(node);
      if (el) results.push(el);
    }

    // Recurse into children
    const children = node.children as unknown[] | undefined;
    if (Array.isArray(children)) {
      collectMatches(children, query, results);
    }
  }
}

/** Walk the tree and return all matching elements. */
export function findElements(
  tree: unknown[],
  selector: string,
): FoundElement[] {
  const query = parseSelector(selector);

  // Collect matches for each OR group, then merge
  const allMatches: FoundElement[] = [];
  const seen = new Set<Record<string, unknown>>();

  for (const group of query) {
    const groupMatches: FoundElement[] = [];
    collectMatches(tree, group, groupMatches);

    // Apply index if present
    if (group.index !== undefined) {
      const idx =
        group.index >= 0
          ? group.index
          : groupMatches.length + group.index;
      if (idx >= 0 && idx < groupMatches.length) {
        const el = groupMatches[idx];
        if (!seen.has(el.node)) {
          seen.add(el.node);
          allMatches.push(el);
        }
      }
    } else {
      for (const el of groupMatches) {
        if (!seen.has(el.node)) {
          seen.add(el.node);
          allMatches.push(el);
        }
      }
    }
  }

  return allMatches;
}

// ---------------------------------------------------------------------------
// ElementHandle — fluent API for interacting with found elements
// ---------------------------------------------------------------------------

import type { ScreenMCPClient } from "./client.js";

export class ElementHandle {
  constructor(
    private client: ScreenMCPClient,
    private selector: string,
    private timeout: number = 3000,
  ) {}

  private async resolve(): Promise<FoundElement> {
    const deadline = Date.now() + this.timeout;
    while (true) {
      const { tree } = await this.client.uiTree();
      const found = findElements(tree, this.selector);
      if (found.length > 0) return found[0];
      if (Date.now() >= deadline) {
        throw new Error(`Element not found: ${this.selector}`);
      }
      await new Promise((r) => setTimeout(r, 500));
    }
  }

  async click(): Promise<void> {
    const el = await this.resolve();
    await this.client.click(el.x, el.y);
  }

  async longClick(): Promise<void> {
    const el = await this.resolve();
    await this.client.longClick(el.x, el.y);
  }

  async type(text: string): Promise<void> {
    const el = await this.resolve();
    await this.client.click(el.x, el.y);
    await new Promise((r) => setTimeout(r, 300));
    await this.client.type(text);
  }

  async element(): Promise<FoundElement> {
    return this.resolve();
  }
}
